File: app/src/main/java/mega/privacy/android/app/TransfersManagementActivity.java
package mega.privacy.android.app;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.widget.RelativeLayout;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import mega.privacy.android.app.components.transferWidget.TransferWidget;
import mega.privacy.android.app.lollipop.ManagerActivityLollipop;
import mega.privacy.android.app.lollipop.PinActivityLollipop;

import static mega.privacy.android.app.components.transferWidget.TransfersManagement.*;
import static mega.privacy.android.app.constants.BroadcastConstants.*;
import static mega.privacy.android.app.lollipop.ManagerActivityLollipop.PENDING_TAB;
import static mega.privacy.android.app.lollipop.ManagerActivityLollipop.TRANSFERS_TAB;
import static mega.privacy.android.app.utils.Constants.*;
import static mega.privacy.android.app.utils.LogUtil.*;

public class TransfersManagementActivity extends PinActivityLollipop {

    protected TransferWidget transfersWidget;

    /**
     * Broadcast to update the transfers widget.
     */
    protected BroadcastReceiver transfersUpdateReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            updateWidget(intent);
        }
    };

    /**
     * Registers the transfers BroadcastReceiver.
     */
    protected void registerTransfersReceiver() {
        registerReceiver(transfersUpdateReceiver, new IntentFilter(BROADCAST_ACTION_INTENT_TRANSFER_UPDATE));
    }

    /**
     * Sets a view as transfers widget.
     *
     * @param transfersWidgetLayout RelativeLayout view to set
     */
    protected void setTransfersWidgetLayout(RelativeLayout transfersWidgetLayout) {
        setTransfersWidgetLayout(transfersWidgetLayout, null);
    }

    /**
     * Sets a view as transfers widget.
     *
     * @param transfersWidgetLayout RelativeLayout view to set
     * @param context               current Context, only used to identify if the view belongs to the ManagerActivity
     */
    protected void setTransfersWidgetLayout(RelativeLayout transfersWidgetLayout, Context context) {
        transfersWidget = new TransferWidget(this, transfersWidgetLayout);
        transfersWidgetLayout.findViewById(R.id.transfers_button).setOnClickListener(v -> {
            if (context instanceof ManagerActivityLollipop) {
                ManagerActivityLollipop.setDrawerItem(ManagerActivityLollipop.DrawerItem.TRANSFERS);
                ((ManagerActivityLollipop) context).selectDrawerItemLollipop(ManagerActivityLollipop.getDrawerItem());
            } else {
                openTransfersSection();
            }
            if (isOnTransferOverQuota()) {
                MegaApplication.getTransfersManagement().setHasNotToBeShowDueToTransferOverQuota(true);
            }
        });
    }

    /**
     * Defines the click action of the transfers widget.
     * Launches an Intent to navigate to In progress tab in Transfers section.
     */
    protected void openTransfersSection() {
        if (megaApi.isLoggedIn() == 0 || dbH.getCredentials() == null) {
            logWarning("No logged in, no action.");
            return;
        }

        Intent intent = new Intent(this, ManagerActivityLollipop.class);
        intent.setAction(ACTION_SHOW_TRANSFERS);
        intent.putExtra(TRANSFERS_TAB, PENDING_TAB);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
        finish();
    }

    /**
     * Updates the state of the transfers widget when the correspondent LocalBroadcast has been received.
     *
     * @param intent    Intent received in the LocalBroadcast
     */
    protected void updateWidget(Intent intent) {
        if (intent == null) return;

        if (transfersWidget != null) {
            int transferType = intent.getIntExtra(TRANSFER_TYPE, INVALID_VALUE);
            if (transferType == INVALID_VALUE) {
                transfersWidget.update();
            } else {
                transfersWidget.update(transferType);
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (transfersWidget != null) {
            transfersWidget.update();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        try {
            //If transfersUpdateReceiver is not registered, it throws an IllegalArgumentException
            unregisterReceiver(transfersUpdateReceiver);
        } catch (IllegalArgumentException e) {
            logWarning("IllegalArgumentException unregistering transfersUpdateReceiver", e);
        }
    }
}

File: app/src/main/java/mega/privacy/android/app/TransfersManagementActivity.kt
package mega.privacy.android.app

import android.content.*
import android.os.Bundle
import android.view.View
import android.widget.RelativeLayout
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.jeremyliao.liveeventbus.LiveEventBus
import mega.privacy.android.app.activities.PasscodeActivity
import mega.privacy.android.app.components.transferWidget.TransferWidget
import mega.privacy.android.app.constants.BroadcastConstants.*
import mega.privacy.android.app.constants.EventConstants.EVENT_SHOW_SCANNING_TRANSFERS_DIALOG
import mega.privacy.android.app.lollipop.DrawerItem
import mega.privacy.android.app.lollipop.ManagerActivity
import mega.privacy.android.app.lollipop.ManagerActivity.PENDING_TAB
import mega.privacy.android.app.lollipop.ManagerActivity.TRANSFERS_TAB
import mega.privacy.android.app.utils.AlertDialogUtil.isAlertDialogShown
import mega.privacy.android.app.utils.Constants.*
import mega.privacy.android.app.utils.LogUtil
import mega.privacy.android.app.utils.StringResourcesUtils
import mega.privacy.android.app.utils.Util

open class TransfersManagementActivity : PasscodeActivity() {

    companion object {
        const val IS_CANCEL_TRANSFERS_SHOWN = "IS_CANCEL_TRANSFERS_SHOWN"
    }

    var transfersWidget: TransferWidget? = null

    private var scanningTransfersDialog: AlertDialog? = null
    private var cancelTransfersDialog: AlertDialog? = null

    /**
     * Broadcast to update the transfers widget when a change in network connection is detected.
     */
    private var networkUpdateReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action != BROADCAST_ACTION_INTENT_CONNECTIVITY_CHANGE) {
                return
            }

            when (intent.getIntExtra(ACTION_TYPE, INVALID_ACTION)) {
                GO_ONLINE -> transfersManagement.resetNetworkTimer()
                GO_OFFLINE -> transfersManagement.startNetworkTimer()
            }
        }
    }

    /**
     * Broadcast to update the transfers widget.
     */
    private var transfersUpdateReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            updateWidget(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupObservers()

        if (savedInstanceState != null) {
            when {
                savedInstanceState.getBoolean(IS_CANCEL_TRANSFERS_SHOWN, false) -> {
                    showCancelTransfersDialog()
                }
                transfersManagement.shouldShowScanningTransfersDialog() -> {
                    showScanningTransfersDialog()
                }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean(IS_CANCEL_TRANSFERS_SHOWN, isAlertDialogShown(cancelTransfersDialog))
        super.onSaveInstanceState(outState)
    }

    /**
     * Registers the transfers BroadcastReceivers and observers.
     */
    private fun setupObservers() {
        registerReceiver(
            transfersUpdateReceiver,
            IntentFilter(BROADCAST_ACTION_INTENT_TRANSFER_UPDATE)
        )

        registerReceiver(
            networkUpdateReceiver,
            IntentFilter(BROADCAST_ACTION_INTENT_CONNECTIVITY_CHANGE)
        )

        LiveEventBus.get(EVENT_SHOW_SCANNING_TRANSFERS_DIALOG, Boolean::class.java)
            .observe(this) { show ->
                when {
                    show && transfersManagement.shouldShowScanningTransfersDialog() -> {
                        showScanningTransfersDialog()
                    }
                    !show && !transfersManagement.shouldShowScanningTransfersDialog() -> {
                        scanningTransfersDialog?.dismiss()
                    }
                }
            }
    }

    /**
     * Sets a view as transfers widget.
     *
     * @param transfersWidgetLayout RelativeLayout view to set
     */
    protected fun setTransfersWidgetLayout(transfersWidgetLayout: RelativeLayout) {
        setTransfersWidgetLayout(transfersWidgetLayout, null)
    }

    /**
     * Sets a view as transfers widget.
     *
     * @param transfersWidgetLayout RelativeLayout view to set
     * @param context               Current Context.
     *                              Only used to identify if the view belongs to the ManagerActivity.
     */
    protected fun setTransfersWidgetLayout(
        transfersWidgetLayout: RelativeLayout,
        context: Context?
    ) {
        transfersWidget = TransferWidget(this, transfersWidgetLayout, transfersManagement)

        transfersWidgetLayout.findViewById<View>(R.id.transfers_button)
            .setOnClickListener {
                if (context is ManagerActivity) {
                    context.drawerItem = DrawerItem.TRANSFERS
                    context.selectDrawerItem(context.drawerItem)
                } else {
                    openTransfersSection()
                }

                if (transfersManagement.isOnTransferOverQuota()) {
                    transfersManagement.setHasNotToBeShowDueToTransferOverQuota(true)
                }
            }
    }

    /**
     * Defines the click action of the transfers widget.
     * Launches an Intent to navigate to In progress tab in Transfers section.
     */
    protected fun openTransfersSection() {
        if (megaApi.isLoggedIn == 0 || dbH.credentials == null) {
            LogUtil.logWarning("No logged in, no action.")
            return
        }

        startActivity(
            Intent(this, ManagerActivity::class.java)
                .setAction(ACTION_SHOW_TRANSFERS)
                .putExtra(TRANSFERS_TAB, PENDING_TAB)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        )

        finish()
    }

    /**
     * Updates the state of the transfers widget when the correspondent LocalBroadcast has been received.
     *
     * @param intent    Intent received in the LocalBroadcast
     */
    protected fun updateWidget(intent: Intent?) {
        if (intent == null || transfersWidget == null) {
            return
        }

        val transferType = intent.getIntExtra(TRANSFER_TYPE, EXTRA_BROADCAST_INVALID_VALUE)

        if (transferType == EXTRA_BROADCAST_INVALID_VALUE) {
            transfersWidget?.update()
        } else {
            transfersWidget?.update(transferType)
        }
    }

    /**
     * Shows a scanning transfers dialog.
     */
    private fun showScanningTransfersDialog() {
        if (isAlertDialogShown(scanningTransfersDialog)) {
            return
        }

        scanningTransfersDialog = MaterialAlertDialogBuilder(this)
            .setView(R.layout.dialog_scanning_transfers)
            .setPositiveButton(
                StringResourcesUtils.getString(R.string.cancel_transfers)
            ) { _, _ ->
                showCancelTransfersDialog()
            }
            .create()
            .apply {
                setCancelable(false)
                setCanceledOnTouchOutside(false)
                show()
            }
    }

    /**
     * Shows a confirmation dialog before cancel all scanning transfers.
     */
    private fun showCancelTransfersDialog() {
        cancelTransfersDialog = MaterialAlertDialogBuilder(this)
            .setTitle(StringResourcesUtils.getString(R.string.cancel_transfers))
            .setMessage(StringResourcesUtils.getString(R.string.warning_cancel_transfers))
            .setPositiveButton(
                StringResourcesUtils.getString(R.string.button_proceed)
            ) { _, _ ->
                transfersManagement.cancelScanningTransfers()
                Util.showSnackbar(this, StringResourcesUtils.getString(R.string.transfers_cancelled))
            }
            .setNegativeButton(StringResourcesUtils.getString(R.string.general_dismiss), null)
            .create()
            .apply {
                setCancelable(false)
                setCanceledOnTouchOutside(false)
                show()
            }
    }

    override fun onResume() {
        super.onResume()

        transfersWidget?.update()
    }

    override fun onDestroy() {
        super.onDestroy()

        unregisterReceiver(transfersUpdateReceiver)
        unregisterReceiver(networkUpdateReceiver)

        scanningTransfersDialog?.dismiss()
        cancelTransfersDialog?.dismiss()
    }

    fun updateWidget() {
        transfersWidget?.update()
    }

    fun hideWidget() {
        transfersWidget?.hide()
    }
}